quinn-proto/src/bytes_source.rs

```diff
@@ -0,0 +1,231 @@
+
+use bytes::Bytes;
+
+/// Indicates how many bytes and chunks had been transferred in a write operation
+#[derive(Debug, Default, PartialEq, Eq, Clone, Copy)]
+pub struct Written {
+    /// The amount of bytes which had been written
+    pub bytes: usize,
+    /// The amount of full chunks which had been written
+    ///
+    /// If a chunk was only partially written, it will not be counted by this field.
+    pub chunks: usize,
+}
+
+/// A source of one or more buffers which can be converted into `Bytes` buffers on demand
+///
+/// The purpose of this data type is to defer conversion as long as possible,
+/// so that no heap allocation is required in case no data is writable.
+pub trait BytesSource {
+    // Returns the total amount of bytes to write
+    fn len(&self) -> usize;
+    /// Limits the source to the new size
+    fn limit(&mut self, new_len: usize);
+    /// The amount of bytes and chunks consumed from this source
+    fn consumed(&self) -> Written;
+    /// Returns the next chunk from the source of owned chunks.
+    ///
+    /// This method will consume parts of the source.
+    // Calling it will yield `Bytes` elements up to the configured `limit`.
+    fn pop_chunk(&mut self) -> Option<Bytes>;
+}
+
+/// A [`BytesSource`] implementation for `&'a mut [Bytes]`
+///
+/// The type allows to dequeue [`Bytes`] chunks from an array of chunks, up to
+/// a configured limit.
+pub struct BytesArray<'a> {
+    /// The wrapped slice of `Bytes`
+    chunks: &'a mut [Bytes],
+    /// The amount of bytes which are still available for consumption
+    len: usize,
+    /// The amount of bytes and chunks consumed from this source
+    consumed: Written,
+}
+
+impl<'a> BytesArray<'a> {
+    pub fn from_chunks(chunks: &'a mut [Bytes]) -> Self {
+        let len = chunks.iter().map(|chunk| chunk.len()).sum();
+        Self {
+            chunks,
+            len,
+            consumed: Written::default(),
+        }
+    }
+}
+
+impl<'a> BytesSource for BytesArray<'a> {
+    fn len(&self) -> usize {
+        self.len
+    }
+
+    fn limit(&mut self, new_len: usize) {
+        self.len = self.len.min(new_len);
+    }
+
+    fn consumed(&self) -> Written {
+        self.consumed
+    }
+
+    fn pop_chunk(&mut self) -> Option<Bytes> {
+        // The loop exists to skip empty chunks while still marking them as
+        // consumed
+        while self.consumed.chunks < self.chunks.len() {
+            let chunk = &mut self.chunks[self.consumed.chunks];
+
+            if chunk.len() <= self.len {
+                let chunk = std::mem::take(chunk);
+                self.len -= chunk.len();
+                self.consumed.chunks += 1;
+                self.consumed.bytes += chunk.len();
+                if chunk.is_empty() {
+                    continue;
+                }
+                return Some(chunk);
+            } else if self.len > 0 {
+                let chunk = chunk.split_to(self.len);
```

Comment:
`split_to()` panics if the argument is greater than the chunk's length, which we don't check here, right?

Reply:
Ah, fair enough, sorry!